### DISCUSSION TITLE
Missing libdl.so and other basic libraries

### DISCUSSION THREAD

**Discussion by Theory177** [dotnet/dotnet-docker#4938] [Q&A] [answered]
While using the new .net 8.0 i got this super weird error, on monday, everything worked fine, last build was on 00:15 gmt +1 (or 2 no idea actually like gets summertime accounted or not?)

Well whatever, the issue is this right here attached as a file is a docker file.

I really have no idea why this suddenly happens, i tried including libraries but i just started using docker this week and i'm kinda baffled to be honest.

Error MSG:

  | Message | "Unable to load shared library 'libdl.so' or one of its dependencies. In order to help diagnose loading problems, consider using a tool like strace. If you're using glibc, consider setting the LD_DEBUG environment variable: \n/usr/share/dotnet/shared/Microsoft.NETCore.App/8.0.0-rc.2.23479.6/libdl.so: cannot open shared object file: No such file or directory\n/app/bin/Debug/net8.0/libdl.so: cannot open shared object file: No such file or directory\n/usr/share/dotnet/shared/Microsoft.NETCore.App/8.0.0-rc.2.23479.6/liblibdl.so: cannot open shared object file: No such file or directory\n/app/bin/Debug/net8.0/liblibdl.so: cannot open shared object file: No such file or directory\n/usr/share/dotnet/shared/Microsoft.NETCore.App/8.0.0-rc.2.23479.6/libdl.so.so: cannot open shared object file: No such file or directory\n/app/bin/Debug/net8.0/libdl.so.so: cannot open shared object file: No such file or directory\n/usr/share/dotnet/shared/Microsoft.NETCore.App/8.0.0-rc.2.23479.6/liblibdl.so.so: cannot open shared object file: No such file or directory\n/app/bin/Debug/net8.0/liblibdl.so.so: cannot open shared object file: No such file or directory\n" | string

As far as i know these should be "standard" libraries but i don't know anything so thats that. In the Dockerfile you see my approach on including those but since i actually don't know anything i'm not sure if this actually would help.

Edit: i should mention i tried searching the Docker Image and they're there. also why are you guys giving us a argument option in the docker compose when it literally only pastes it as a string or opens a script and prints the lines as text (just curious here bc this really grinds my gears)? 

Edit2: Also to mention i'm not sure if this is smth Microsoft related or actually from the library i use for pdfs, anyway i take every straw so i posted here aswell, the original post is: https://forum.gemboxsoftware.com/t/missing-library-on-dockerized-app/1210

Edit3: This is probably the best thing i found from all around the www https://stackoverflow.com/questions/75855053/how-to-address-crash-due-to-missing-libdl-so-on-ubuntu-22 
What i also forgot to mention is i switched from VS Professional to Community bc of trial ending :) 

[Dockerfile.txt](https://github.com/dotnet/dotnet-docker/files/12864409/Dockerfile.txt)

Kindly K

**Comment by timmi-on-rails** [comment]
Same problem here and I am using PdfSharp https://github.com/empira/PDFsharp
Did you find a solution?

**Comment by Steve887** [comment]
I solved this by linking `libdl.so.2` to `libdl.so`. You can use `whereis libdl.so.2` to confirm where the files are if they are in a different location.

    RUN ln -s /usr/lib/x86_64-linux-gnu/libdl.so.2 /usr/lib/x86_64-linux-gnu/libdl.so

See this SO answer: https://unix.stackexchange.com/a/709036

**Reply by scalalang2** [reply]
This works for me :) thx

**Reply by iheartcsharp** [reply]
Worked for me as well.

**Reply by ben-aslan** [reply]
Worked for me too as well     tnx

**Comment by EvilBeaver** [comment]
This answer https://stackoverflow.com/questions/75855053/how-to-address-crash-due-to-missing-libdl-so-on-ubuntu-22 says: 
> This is a building issue with the application, the application was incorrectly built and linked against libdl.so. For Linux support, the software must be rebuilt and linked to libdl.so.2 instead.

Does this mean Microsoft did a wrong build of dotnet?

**Comment by jkotas** [comment, marked as answer]
> Unable to load shared library 'libdl.so' or one of its dependencies. In order to help diagnose loading problems, consider using a tool like strace.

The error with this specific error message is caused by some application or library code trying to PInvoke `libdl.so` directly, like this: https://github.com/Therzok/SQLitePCL.raw/blob/d86a5b5311ff4cdbba54cda49c23a776bd888663/src/cs/sqlite3_pinvoke.cs#L1255-L1257

The proper fix is change the offending code to call [NativeLibrary.Load](https://learn.microsoft.com/dotnet/api/system.runtime.interopservices.nativelibrary.load) managed API that abstracts away the platforms and platform version differences in native library loading.

Linking `libdl.so` to `libdl.so.2` is a fine workaround until the proper fix can be deployed.

**Comment by jkotas** [comment]
cc @dotnet/interop-contrib for awareness.

**Comment by khaledmahmoud78** [comment]
@Steve887 It works! ... Thanks a lot!

**Comment by xanth** [comment]
Related issue observed in [confluent-kafka-dotnet](https://github.com/confluentinc/confluent-kafka-dotnet) - [Unable to load librdkafka.so due to direct dependancy on libdl #2550](https://github.com/confluentinc/confluent-kafka-dotnet/issues/2550)
